From a2d1076f7348930a7c0dc1278f36275bce235032 Mon Sep 17 00:00:00 2001
From: Hanlu Li <heuleehanlu@gmail.com>
Date: Thu, 17 Apr 2025 11:21:08 +0800
Subject: [PATCH 1046/1058] LATX, fix: Add fallback for vector initialization
 using raw instruction encoding

Some system use GCC versions that do not support LoongArch vector extensions such
as LSX or LASX. To improve compatibility, replace vector register initialization
with equivalent raw instruction encoding using inline assembly.

Signed-off-by: Hanlu Li <heuleehanlu@gmail.com>
---
 linux-user/i386/signal.c       |   1 -
 target/i386/latx/latx-config.c | 131 +++++++++++++++++----------------
 2 files changed, 66 insertions(+), 66 deletions(-)

diff --git a/linux-user/i386/signal.c b/linux-user/i386/signal.c
index cdec4ccf07..280ee084ba 100644
--- a/linux-user/i386/signal.c
+++ b/linux-user/i386/signal.c
@@ -23,7 +23,6 @@
 #ifdef CONFIG_LATX
 #include "latx-options.h"
 #include "reg-map.h"
-#include <lasxintrin.h>
 #endif
 
 /* from the Linux kernel - /arch/x86/include/uapi/asm/sigcontext.h */
diff --git a/target/i386/latx/latx-config.c b/target/i386/latx/latx-config.c
index 83a06eeaea..eca0f6114a 100644
--- a/target/i386/latx/latx-config.c
+++ b/target/i386/latx/latx-config.c
@@ -502,79 +502,80 @@ void latx_dt_init(void)
 #endif
 }
 
-#include <lasxintrin.h>
 void latx_init_fpu_regs(CPUArchState *env)
 {
     if (option_enable_lasx) {
+        /* vldi $vr0,0 */
         asm volatile (
-            "xvldi $xr0,0\r\n"
-            "xvldi $xr1,0\r\n"
-            "xvldi $xr2,0\r\n"
-            "xvldi $xr3,0\r\n"
-            "xvldi $xr4,0\r\n"
-            "xvldi $xr5,0\r\n"
-            "xvldi $xr6,0\r\n"
-            "xvldi $xr7,0\r\n"
-            "xvldi $xr8,0\r\n"
-            "xvldi $xr9,0\r\n"
-            "xvldi $xr10,0\r\n"
-            "xvldi $xr11,0\r\n"
-            "xvldi $xr12,0\r\n"
-            "xvldi $xr13,0\r\n"
-            "xvldi $xr14,0\r\n"
-            "xvldi $xr15,0\r\n"
-            "xvldi $xr16,0\r\n"
-            "xvldi $xr17,0\r\n"
-            "xvldi $xr18,0\r\n"
-            "xvldi $xr19,0\r\n"
-            "xvldi $xr20,0\r\n"
-            "xvldi $xr21,0\r\n"
-            "xvldi $xr22,0\r\n"
-            "xvldi $xr23,0\r\n"
-            "xvldi $xr24,0\r\n"
-            "xvldi $xr25,0\r\n"
-            "xvldi $xr26,0\r\n"
-            "xvldi $xr27,0\r\n"
-            "xvldi $xr28,0\r\n"
-            "xvldi $xr29,0\r\n"
-            "xvldi $xr30,0\r\n"
-            "xvldi $xr31,0\r\n"
+            ".word (0x77e00000 | (0 << 5) | (0)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (1)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (2)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (3)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (4)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (5)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (6)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (7)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (8)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (9)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (10)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (11)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (12)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (13)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (14)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (15)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (16)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (17)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (18)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (19)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (20)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (21)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (22)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (23)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (24)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (25)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (26)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (27)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (28)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (29)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (30)) \r\n"
+            ".word (0x77e00000 | (0 << 5) | (31)) \r\n"
             : : :
         );
     } else {
+        /* vldi $vr0,0 */
         asm volatile (
-            "vldi $vr0,0\r\n"
-            "vldi $vr1,0\r\n"
-            "vldi $vr2,0\r\n"
-            "vldi $vr3,0\r\n"
-            "vldi $vr4,0\r\n"
-            "vldi $vr5,0\r\n"
-            "vldi $vr6,0\r\n"
-            "vldi $vr7,0\r\n"
-            "vldi $vr8,0\r\n"
-            "vldi $vr9,0\r\n"
-            "vldi $vr10,0\r\n"
-            "vldi $vr11,0\r\n"
-            "vldi $vr12,0\r\n"
-            "vldi $vr13,0\r\n"
-            "vldi $vr14,0\r\n"
-            "vldi $vr15,0\r\n"
-            "vldi $vr16,0\r\n"
-            "vldi $vr17,0\r\n"
-            "vldi $vr18,0\r\n"
-            "vldi $vr19,0\r\n"
-            "vldi $vr20,0\r\n"
-            "vldi $vr21,0\r\n"
-            "vldi $vr22,0\r\n"
-            "vldi $vr23,0\r\n"
-            "vldi $vr24,0\r\n"
-            "vldi $vr25,0\r\n"
-            "vldi $vr26,0\r\n"
-            "vldi $vr27,0\r\n"
-            "vldi $vr28,0\r\n"
-            "vldi $vr29,0\r\n"
-            "vldi $vr30,0\r\n"
-            "vldi $vr31,0\r\n"
+            ".word (0x73e00000 | (0 << 5) | (0)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (1)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (2)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (3)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (4)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (5)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (6)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (7)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (8)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (9)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (10)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (11)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (12)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (13)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (14)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (15)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (16)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (17)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (18)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (19)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (20)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (21)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (22)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (23)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (24)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (25)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (26)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (27)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (28)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (29)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (30)) \r\n"
+            ".word (0x73e00000 | (0 << 5) | (31)) \r\n"
             : : :
         );
     }
-- 
2.49.0

